crates/bevy_ui/src/geometry.rs

```diff
@@ -45,6 +55,70 @@ pub enum Val {
     VMax(f32),
 }
 
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum ValParseError {
+    UnitMissing,
+    ValueMissing,
+    InvalidValue,
+    InvalidUnit,
+}
+
+impl core::fmt::Display for ValParseError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            ValParseError::UnitMissing => write!(f, "unit missing"),
+            ValParseError::ValueMissing => write!(f, "value missing"),
+            ValParseError::InvalidValue => write!(f, "invalid value"),
+            ValParseError::InvalidUnit => write!(f, "invalid unit"),
+        }
+    }
+}
+
+impl core::str::FromStr for Val {
+    type Err = ValParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let s = s.trim();
+
+        if s.eq_ignore_ascii_case("auto") {
+            return Ok(Val::Auto);
+        }
+
+        let Some(end_of_number) = s
+            .bytes()
+            .position(|c| !(c.is_ascii_digit() || c == b'.' || c == b'-' || c == b'+'))
+        else {
+            return Err(ValParseError::UnitMissing);
+        };
+
+        if end_of_number == 0 {
+            return Err(ValParseError::ValueMissing);
+        }
+
+        let (value, unit) = s.split_at(end_of_number);
+
+        let value: f32 = value.parse().map_err(|_| ValParseError::InvalidValue)?;
+
+        let unit = unit.trim();
+
+        if unit.eq_ignore_ascii_case("px") {
+            Ok(Val::Px(value))
+        } else if unit.eq_ignore_ascii_case("%") {
+            Ok(Val::Percent(value))
+        } else if unit.eq_ignore_ascii_case("vw") {
+            Ok(Val::Vw(value))
+        } else if unit.eq_ignore_ascii_case("vh") {
+            Ok(Val::Vh(value))
+        } else if unit.eq_ignore_ascii_case("vmin") {
+            Ok(Val::VMin(value))
+        } else if unit.eq_ignore_ascii_case("vmax") {
+            Ok(Val::VMax(value))
+        } else {
+            Err(ValParseError::InvalidUnit)
+        }
```

Comment:
The long `else if` chain is a bit ugly, is there a neater construction we could use?

Maybe something like:
```suggestion
        const VALS: &[(&str, fn(f32) -> Val)] = &[
            ("px", Val::Px),
            ("%", Val::Percent),
            ("vw", Val::Vw),
            ("vh", Val::Vh),
            ("vmin", Val::VMin),
            ("vmax", Val::VMax),
        ];

        for (name, val) in VALS {
            if unit.eq_ignore_ascii_case(name) {
                return Ok(val(value));
            }
        }

        Err(ValParseError::InvalidUnit)
```

Or less nice but non-looping:
```
        let val = if unit.eq_ignore_ascii_case("px") {
            Val::Px
        } else if unit.eq_ignore_ascii_case("%") {
            Val::Percent
        } else if unit.eq_ignore_ascii_case("vw") {
            Val::Vw
        } else if unit.eq_ignore_ascii_case("vh") {
            Val::Vh
        } else if unit.eq_ignore_ascii_case("vmin") {
            Val::VMin
        } else if unit.eq_ignore_ascii_case("vmax") {
            Val::VMax
        } else {
            return Err(ValParseError::InvalidUnit);
        };

        Ok(val(value))
```

This isn't that important though, and otherwise everything looks good.

Reply:
I think it's fine how it is. Your suggestion might be nicer if it didn't require a loop, but the way it is seems the most straightforward and readable.